jitrebalance/jitrebalance.py

```diff
@@ -0,0 +1,184 @@
+            "payment_hash": payment_hash,
+            "request": request,
+        }
+
+        # After all this work we're finally in a position to judge whether a
+        # rebalancing is worth it at all. The rebalancing is considered worth it
+        # if the fees we're about to pay are less than or equal to the fees we get
+        # out of forwarding the payment.
+        plugin.log("Sending rebalance request using payment_hash={}, route={}".format(
+            payment_hash, route
+        ))
+        try:
+            plugin.rpc.sendpay(route, payment_hash)
+            # If the attempt is successful, we acknowledged it on the
+            # receiving end (a couple of line above), so we leave it dangling
+            # here.
+            plugin.rpc.waitsendpay(payment_hash)
+            return
+        except RpcError as e:
+            error = e.error['data']
+            erring_channel = error['erring_channel']
+            exclusions.append(erring_channel)
+            plugin.log("Excluding {} due to a failed attempt".format(erring_channel))
+
+    request.set_result({"result": "continue"})
+
+
+@plugin.async_hook("htlc_accepted")
+def on_htlc_accepted(htlc, onion, plugin, request, **kwargs):
+    plugin.log("Got an incoming HTLC htlc={}, onion={}".format(htlc, onion))
+
+    # The HTLC might be a rebalance we ourselves initiated, better check
+    # against the list of pending ones.
+    rebalance = plugin.rebalances.get(htlc['payment_hash'], None)
+    if rebalance is not None:
+        # Settle the rebalance, before settling the request that initiated the
+        # rebalance.
+        request.set_result({
+            "result": "resolve",
+            "payment_key": rebalance['payment_key']
+        })
+
+        # Now wait for it to settle correctly
+        # TODO Maybe be a bit smarter than having a fixed timeout here?
+        time.sleep(1)
+
+        rebalance['request'].set_result({"result": "continue"})
+
+        # Clean up our stash of active rebalancings.
+        del plugin.rebalances[htlc['payment_hash']]
+        return
+
+    # Check to see if the next channel has sufficient capacity
+    scid = onion['short_channel_id'] if 'short_channel_id' in onion else '0x0x0'
+
+    # Are we the destination? Then there's nothing to do. Continue.
+    if scid == '0x0x0':
+        request.set_result({"result": "continue"})
+        return
+
+    # Locate the channel + direction that would be the next in the path
+    peers = plugin.rpc.listpeers()['peers']
+
+    chan = None
+    peer = None
+    for p in peers:
+        for c in p['channels']:
+            if c['short_channel_id'] == scid:
+                chan = c
+                peer = p
+
+    # Check if the channel is active and routable, otherwise there's little
+    # point in even trying
+    if not peer['connected'] or chan['state'] != "CHANNELD_NORMAL":
+        request.set_result({"result": "continue"})
```

Comment:
Aside: Could we test using another outgoing channel instead ? I'm planning to do a probe-then-forward plugin and that would be the exact case.

Reply:
Uhm, we can't really redirect an HTLC to another node (and we don't have multi-channels with a single peer) since the onion is only decodeable by the correct next peer.

But we can probe and do any shenannigans we want before forwarding :-)